pkg/devfile/devfile.go

```diff
@@ -217,26 +217,27 @@ func DownloadDevfile(dir string) ([]byte, error) {
 		devfileBytes, err = DownloadFile(devfilePath)
 		if err == nil {
 			// if we get a 200, return
-			return devfileBytes, err
+			return devfileBytes, path, err
 		}
 	}
 
-	return nil, &NoDevfileFound{Location: dir}
+	return nil, "", &NoDevfileFound{Location: dir}
 }
 
 // DownloadFile downloads the specified file
 func DownloadFile(file string) ([]byte, error) {
 	return util.CurlEndpoint(file)
 }
 
-// DownloadDevfileAndDockerfile attempts to download the  devfile and dockerfile from the root of the specified url
-func DownloadDevfileAndDockerfile(url string) ([]byte, []byte) {
+// DownloadDevfileAndDockerfile attempts to download and return the devfile, devfile context and dockerfile from the root of the specified url
+func DownloadDevfileAndDockerfile(url string) ([]byte, string, []byte) {
```

Comment:
yeah they can be different, the context we pass in for a component. We assume devfile and dockerfile in this context but it could be somewhere else. This is when we do the detection for devfile and dockerfile for upto 1 nested directory. If we find both devfile and dockerfile, well and good. If we find only a dockerfile its a dockerfile component. If we find devfile but not a dockerfile we check if the dockerfile is defined in the devfile otherwise we use alizer to match a component from our devfile sample and use that dockerfile. Thats the gist of it.

Reply:
So when this function is being called, it checks for the devfile at the context level. If we dont find it here, we walk the filepath later to see other directories.